### DISCUSSION TITLE
Packages/com.unity.render-pipelines.universal/manual/anti-aliasing

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#578] [Notes]
# Packages/com.unity.render-pipelines.universal/manual/anti-aliasing

https://docs.unity3d.com/Packages/com.unity.render-pipelines.universal@latest/index.html?subfolder=%2Fmanual%2Fanti-aliasing.html

**Comment by radiatoryang** [comment]
For more on MSAA's cost in 2022, see https://forum.unity.com/threads/msaa-performance-on-mobile.825450/

> MSAA gives quality antialiasing, while being relatively cheap for forward rendering.
With additional samples per pixel, rasterization results in smoother edges.
The additional samples increase memory usage (larger buffer to contain the additional samples).
x4 would require more sub samples per pixel, thus more memory usage as well as more expensive resolve.

> This is also the reason why MSAA is less suited for Deferred Rendering. (Due to multiple GBuffer attachments / increased memory usage)

> MSAA has to be resolved (downsampled) regardless of postprocessing, before display. By "resolving", this simply means averaging the subsamples per pixel. This is done either with a blit/copy operation, or fullscreen shader pass.

> When and how resolve happens can vary, and can impact performance. If post processing effects need to read/write multiple samples per fragment - this can be quite expensive. Thus resolve is often performed before post-processing.

> In the case of URP, a "standard" resolve is applied using a blit/copy operation to a single-sample buffer. This operation corresponds to 'RenderTexture.ResolveAA' sample seen in the frame debugger.

> On modern APIs (Vulkan, Metal), we can specify a resolve attachment in the render pass description, and resolve is handled accordingly. The attachment contents are cleared after the render pass, or stored in memory based on the store operations. (Store, Resolve, StoreAndResolve, DontCare)

> To take advantage of render-target optimizations for tile-based GPUs, as Aleksander mentioned above, set the attachment store operation as DontCare when possible.

> Edit: To clarify, MSAA is definitely not free! as Aleksander mentioned in the first post, cost can be significant if we store the multiplesampled attachment in main memory, so avoid using MSAA attachments with Store or StoreAndResolve.

> One noticeable example is when enabling URPs 'forced depth prepass' with MSAA, which will have a performance impact as we store the multisampled depth attachment, and thus should be avoided.
